Pioneer: Applying for storage working group opening and final "Working groups" tab issues

Lezek123 opened the pull request:
Implements: https://github.com/Joystream/joystream/issues/597

Note that the listing of user's active applications for storage working group etc. is not implemented yet, as there is another issue for that (https://github.com/Joystream/joystream/issues/618)

**ts-ignore issue**
While working on this issue I removed the `ts-ignore` inside `/joy-roles/src/flows/apply.controller.tsx` as it is a very bad practice to use it, because it silences a lot of potential TypeScript errors during the build. It should have actually been caught by eslinter's `ban-ts-ignore` rule, but for some reason it wasn't (perhaps because it wasn't added as a valid-syntax comment, but actually a string value inside jsx).

This forced me to fix a few typing errors inside `/joy-roles/src/flows/apply.tsx` that were revealed once the `ts-ignore` was removed.

**Other pre-existing UI issues**
I also decided to list all the pre-existing `Working Group` tab issues I was able to find in https://github.com/Joystream/joystream/issues/741 to allow better understanding of the problems that may occur during testing, but were not actually introduced in this PR (I'll try to tackle those once the `Working Groups` tab is fully adjusted to support Storage Working Group)

Commit:
Applying for storage working group opening

Commit:
Pioneer: Storage working group "My applicatons" and "My roles"

Commit:
Linter issues fix

Commit:
Opportunities: Neverneding loading screen fix

Commit:
Apply flow - stakes fixes

Commit:
Role key naming fix

Lezek123 commented:
Updated to include https://github.com/Joystream/joystream/issues/618 and some of the fixes from https://github.com/Joystream/joystream/issues/741:

`working-groups/my-roles` page should now be fully adjusted to work with Storage Working Group. Although in issue https://github.com/Joystream/joystream/issues/618 it wasn't planned to make it fully compatible (active roles were not considered), I decided to go for it, since it was very easy to implement at this point and makes the Pioneer's support for Storage Working Group _complete_.

Issues described in https://github.com/Joystream/joystream/issues/741:
- Fixed the issue with neverending loading screen in case there are no openings available.
- Fixed the applying flow for openings with variable stakes and max. number of applicants specified. User should now be able to correcty determine the minimum stake he needs to specify to be considered when applying for such openings, as well as his position in the "applicants hierarchy" during the first step of the application. Note that if both stakes are set to `AtLeast`, a different component handles this step than when only one of them is set to `AtLeast`.
- Fixed role key name generation issue (keys with duplicate names should not be generated anymore)

Commit:
Merge branch 'nicaea-types-2nd-update' into storage-working-group-applying

Commit:
Merge branch 'nicaea-types-2nd-update' into storage-working-group-applying

Commit:
Updates following runtime renaming

Commit:
Merge branch 'nicaea-types-2nd-update' into storage-working-group-applying

Commit:
Merge branch 'nicaea' into storage-working-group-applying

mnaamani commented on pioneer/packages/joy-roles/src/tabs/MyRoles.tsx:
```diff
@@ -102,9 +105,10 @@ function RoleName (props: NameAndURL) {
 
 export interface ActiveRole extends NameAndURL {
-  curatorId: CuratorId;
+  workerId: CuratorId;
```
Maybe start to transition to the equivalent types from working-group ?

Lezek123 replied:
I overlooked that one, fixed.

mnaamani commented on pioneer/packages/joy-roles/src/transport.substrate.ts:
```diff
@@ -727,13 +760,13 @@ export class Transport extends TransportBase implements ITransport {
-            applicationText
+          const tx = this.apiExtrinsicByGroup(group, 'applyOnOpening')(
+            membershipIds[0], // Member id
+            new u32(id), // Worker/Curator opening id
```
opening id in the runtime is actually a u64. I think just passing a number is fine, the type will be converted to the correct type expected by the transaction. I realize this code is from prior iterations but I just spotted it now. Using the wrong type will affect the remaining values in the encoded transaction. Its probably best to use `WorkerId` here.

mnaamani replied:
maybe we should check other places where this mistake might have been made.

Lezek123 replied:
Fixed. There were some occurences like that in `Admin.controller.tsx` too. Those weren't causing any trouble porobably because `@polkadot/api` knows how to convert any input according to the api metadata (unless the id is > 4294967295 in which case the`new u32()` would be a problem), but it's good to keep those consistent.

mnaamani reviewed (changes requested):
I left a comment about using correct type in one of the tx calls.

Tested the UI. Only thing I noticed that is incorrect (perhaps its existing behavior) but when appliccations are withdrawn, the total number of applications on an opening still counts the withdrawn applications.

Otherwise all works as expected.

Commit:
Review comments + fix "ts-nocheck" in Admin controller

Lezek123 commented:
Added fixes according to the comments.

I noticed that there was a `ts-nocheck` clause inside `Admin.controller.tsx` which silenced a lot of useful TypeScript errors. I fixed those and it seems to have simultaneously fixed a few UI issues with the opening creation form at `working-groups/admin` (ie. there were some problems with changing the stake type between `AtLeast` and `Exact`, changing stake values etc. and they seem to be gone now).

I added `ban-ts-comment` rule to eslint to possibly avoid running into this problem again. There is one more occurrence of `ts-ignore` in `/packages/react-components/src/AddressCard.tsx` (which was purposefully silenced, so I just updated the `eslint-disable-next-line` clause there) and 2 `ts-nocheck` occurrences in storybook-related files inside `joy-roles` (which I silenced for now with `eslint-disable-next-line`, but marked with `TODO: FIXME:`)

Lezek123 commented:
> Only thing I noticed that is incorrect (perhaps its existing behavior) but when appliccations are withdrawn, the total number of applications on an opening still counts the withdrawn applications.

I don't think I was chaning anything there, it looks like this number just represents total number of applications regardless of their current status (which can be `Active`, `Unstaking` or `Inactive`). It is also displayed for already filled openings etc., so it makes some sense, but I understand that pairing it with `max_active_applications` can be confusing and indicate that this number means _active_ applications.

Commit:
Fix non-active applications included in ranks calculation

Lezek123 commented:
I actually found another issue related to those non-active applications.

Because the status of them wasn't checked inside `myApplicationRank` and `openingApplicationRanks` methods, it caused unreliable rank perdiction and incorrect `minStake` value when applying in case one of the previous applicants has withdrawn its application.

Should be fixed now.

Commit:
Working group overview - separate lead from workers

Commit:
Openings - filter by group lead

Commit:
Final UI adjustments related to leader-worker differentiation

Commit:
Fix build/lint errors

mnaamani commented:
I created an opening with an empty `{}` json for the human readable text so when I visit the `my-roles` page it crashes with the following error (I presume its due to not finding the expected properties on the parsed object):

![Screen Shot 2020-07-07 at 10 25 42 AM](https://user-images.githubusercontent.com/1621012/86737030-05c08180-c045-11ea-99df-1b2c3dbac19e.png)

Maybe we should handle badly formatted human readable text.

Commit:
Merge nicaea

Commit:
Parse HRT fallback

Commit:
Merge branch 'storage-working-group-applying-updated' into storage-working-group-applying

Commit:
Updated to use "JoyEnum"

Lezek123 commented:
Updated with some UI adjustments following "Make lead hireable" changes (https://github.com/Joystream/joystream/issues/726) that I already had on my branch locally.

Also merged `nicaea` to fix conflicts and added a method to parse `human_readable_text` with fallback to valid `GenericJoyStreamRoleSchema` object that has all required values set to `Unknown` in case the actual json is invalid.

mnaamani commented on pioneer/.eslintrc.js:
```diff
@@ -15,6 +15,7 @@ module.exports = {
     'new-cap': 'off',
-    '@typescript-eslint/interface-name-prefix': 'off'
+    '@typescript-eslint/interface-name-prefix': 'off',
+    '@typescript-eslint/ban-ts-comment': 'error'
```
This is a good rule to have:

` '@typescript-eslint/ban-ts-comment': 'error'`

might add it to #891 cc: @Gamaranto @kdembler

kdembler replied:
That's a good one in general but I would use the option `allow-with-description` so we can use `ts-ignore` in specific cases. Recently I've worked with a library with stale typings and had to ignore typescript error to use some newer option.

mnaamani commented on pioneer/packages/joy-roles/src/flows/apply.elements.stories.tsx:
```diff
@@ -1,3 +1,5 @@
+// TODO: FIXME: Remove the ts-nocheck and fix errors!
```
It better to create github issues rather than TODO's in the code. We can label them with `good first issue` for contributors to find something simple and easy to work on.

mnaamani commented on pioneer/packages/joy-roles/src/tabs/Admin.controller.tsx:
```diff
@@ -553,7 +399,9 @@ export class AdminController extends Controller<State, ITransport> {
-    const tx = this.api.tx.contentWorkingGroup.acceptCuratorApplications(new u32(id));
+    const tx = this.api.tx.contentWorkingGroup.acceptCuratorApplications(id);
+    // FIXME: That's a bad way to send extrinsic in Pioneer (without "queueExtrinsic" etc.)
+    // and probably the reason why it always appears as succesful
```
I'm sure here you mean that you are not handling case when the transaction returns an error, but is still finalized?

mnaamani commented:
Works as expected.
One enhancement that I would suggest is in the my-roles, to show in the current roles, the worker id.
This is needed by the storage providers when running their node.
